server/sonar-server/src/main/java/org/sonar/server/authentication/UserIdentityAuthenticator.java

```diff
@@ -56,34 +63,78 @@ private long register(UserIdentity user, IdentityProvider provider) {
+    if (!provider.allowsUsersToSignUp()) {
+      throw new UnauthorizedException(format("'%s' users are not allowed to sign up", provider.getKey()));
+    }
 
-      String email = user.getEmail();
-      if (email != null && dbClient.userDao().doesEmailExist(dbSession, email)) {
-        throw new UnauthorizedException(format(
-          "You can't sign up because email '%s' is already used by an existing user. This means that you probably already registered with another account.", email));
-      }
+    String email = user.getEmail();
+    if (email != null && dbClient.userDao().doesEmailExist(dbSession, email)) {
+      throw new UnauthorizedException(format(
+        "You can't sign up because email '%s' is already used by an existing user. This means that you probably already registered with another account.", email));
+    }
 
-      userUpdater.create(dbSession, NewUser.create()
-        .setLogin(userLogin)
-        .setEmail(user.getEmail())
-        .setName(user.getName())
-        .setExternalIdentity(new ExternalIdentity(provider.getKey(), user.getProviderLogin()))
-        );
-      return dbClient.userDao().selectOrFailByLogin(dbSession, userLogin).getId();
+    String userLogin = user.getLogin();
+    userUpdater.create(dbSession, NewUser.create()
+      .setLogin(userLogin)
+      .setEmail(user.getEmail())
+      .setName(user.getName())
+      .setExternalIdentity(new ExternalIdentity(provider.getKey(), user.getProviderLogin()))
+      );
+    UserDto newUser = dbClient.userDao().selectOrFailByLogin(dbSession, userLogin);
+    if (user.shouldSyncGroups()) {
+      addGroups(dbSession, newUser, user.getGroups());
+    }
+    return newUser.getId();
+  }
 
-    } finally {
-      dbClient.closeSession(dbSession);
+  private void registerExistingUser(DbSession dbSession, UserDto userDto, UserIdentity user, IdentityProvider provider) {
+    userUpdater.update(dbSession, UpdateUser.create(userDto.getLogin())
+      .setEmail(user.getEmail())
+      .setName(user.getName())
+      .setExternalIdentity(new ExternalIdentity(provider.getKey(), user.getProviderLogin()))
+      .setPassword(null));
+    if (user.shouldSyncGroups()) {
+      syncGroups(dbSession, user, userDto);
+    }
+  }
+
+  private void syncGroups(DbSession dbSession, UserIdentity userIdentity, UserDto userDto) {
+    String userLogin = userIdentity.getLogin();
+    Set<String> userGroups = new HashSet<>(dbClient.groupMembershipDao().selectGroupsByLogins(dbSession, singletonList(userLogin)).get(userLogin));
+    Set<String> identityGroups = userIdentity.getGroups();
+
+    Collection<String> groupsToAdd = Sets.difference(identityGroups, userGroups);
+    addGroups(dbSession, userDto, groupsToAdd);
+
+    Collection<String> groupsToRemove = Sets.difference(userGroups, identityGroups);
+    removeGroups(dbSession, userDto, groupsToRemove);
+  }
+
+  private void addGroups(DbSession dbSession, UserDto userDto, Collection<String> groups) {
+    if (!groups.isEmpty()) {
```

Comment:
There's already a UT to check that => authenticate_existing_user_and_remove_all_groups